### ISSUE TITLE
wasm2c use of globals and statics introduce race conditions in concurrent environments

### ISSUE THREAD

**Issue by shravanrn** [WebAssembly/wabt#2046]
wasm2c has a number of globals and statics that assume that wasm instances are only embedded in a single threaded host. 

However, when using wasm2c in complex hosts like Firefox, different host threads can spin up multiple sandbox instances concurrently, invoke function in their respective sandboxes and destroy sandboxes. In this use case, a number of these globals and statics can lead to exploitable race conditions --- hence in the fork of wasm2c used in Firefox, we have completely eliminated the use of such global state.

Some of the current uses of global state in order of security severity are

- [wasm_rt_jmp_buf](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L67) which is used in recovering from traps and if exceptions are enabled
- [wasm_rt_call_stack_depth/wasm_rt_saved_call_stack_depth](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L60-L61) which is used if stack depth counting is employed on the target platform
- [wasm exception related state](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L69-L73) which is used if exceptions are enabled
- [s_tag_count](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L136) which is used to allocate tag numbers for exceptions
- [g_func_types/g_func_type_count](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L64-L65) which is used to give unique type indexes to Wasm indirection function types

I got part way into solving this, however I don't have the cycles to completely address at the moment.

**Comment by shravanrn** [comment]
From what I can tell, the above list contains two different types of state
- Module specific state
- Thread specific state

**Module specific state**
Module specific state from the above would be the [g_func_types/g_func_type_count](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L64-L65)

Here is my attempt at solving this
https://github.com/WebAssembly/wabt/compare/main...PLSysSec:wabt:remove-race-conditions-module

The approach here is to create a new structure `struct wasm_rt_module_state` that is associated with the every wasm instance struct. This struct holds the func_types and is passed into [wasm_rt_register_func_type](https://github.com/WebAssembly/wabt/compare/main...PLSysSec:wabt:remove-race-conditions-module#diff-4f128b661d343d97ee188574413f9fefe659d761a04752fd2970eb9c437618a8R258) so that func_types of each module are isolated. This means each module can be initialized in parallel without any issues to this data

This solution assumes that different modules do not need the same type for indirect function calls

**Thread specific state**

Other variables like 
[wasm_rt_jmp_buf](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L67) [wasm_rt_call_stack_depth/wasm_rt_saved_call_stack_depth](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L60-L61) [wasm exception related state](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L69-L73)  can be thought of as thread specific state that is part of the currently executing Wasm thread. We can thus create a new `struct wasm_rt_thread_state` for this data.

This struct is then passed in either
1. as an extra function parameter to wasm2c functions (similar to the instance struct as of today) [See here](https://github.com/WebAssembly/wabt/compare/main...PLSysSec:wabt:remove-race-conditions-threadparam). This approach is clean but likely has some performance hit due to passing around extra parameters in every function
2. stored in a thread local variable prior to execution [See here](https://github.com/WebAssembly/wabt/compare/main...PLSysSec:wabt:remove-race-conditions-tlsvar). This is probably more efficient, but the implementation must make sure to correctly set up the thread local variable prior to executing wasm2c functions, so it is more prone to error

**Comment by shravanrn** [comment]
fyi @sbc100 @talg @keithw @kripken

**Comment by sbc100** [comment]
Makes sense to me.   I guess we also have "instance specific state" but that is already captured by the recent instancing changes.   Makes sense to me that and instance would have a pointer to the module state.

**Comment by sbc100** [comment]
Looking again, I'm not sure that g_func_types/g_func_type_count really are module specific... at least they seems like could equally well be instance-specific, or actually global to the entire runtime.  

I assume that you could also fix the issue by making access to g_func_types thread safe?   Since g_func_types/g_func_type_count  are only using a single function (wasm_rt_register_func_type) which is only used during module initialization, perhaps we can just protect them locks?

**Comment by sbc100** [comment]
(What is more the lock would only be needed when adding *new* types, not when registering existing types, so the number of calls wasm_rt_register_func_type that need to lock at all, would approach zero over time).

**Comment by shravanrn** [comment]
> I'm not sure that g_func_types/g_func_type_count really are module specific... at least they seems like could equally well be instance-specific, or actually global to the entire runtime.

Yup, In our wasm2c fork, we have this as part of instance state. This does mean that you have to duplicate this work for each instance of a particular module that you spin up, but that is not usually a bottleneck for our use. I am happy to have this as instance state if it simplifies things. 

>  g_func_types thread safe ... perhaps we can just protect them locks? (What is more the lock would only be needed when adding new types, not when registering existing types

We should be a little careful here. We either need reader writer locks or we would need to use a writer only lock as well as an atomic count. I am a little weary of being too clever here, as the weaker memory models on ARM machines usually tend to cause issues if we have the slightest bug. Assuming we do something straightforward, this will come with a perf cost --- I can probably say that this cost would be minor for the Firefox use case, so this approach would be fine too!

**Comment by shravanrn** [comment]
oh also for what its worth, i spotted a spec test that requires exception state to be shared across modules that interact with each other... so this state would definitely have to be handled separately from instance state

**Comment by keithw** [comment]
@shravanrn Thanks for looping me in! In general I do think it's valuable to maintain the API abstraction and avoid treating wasm-rt-impl.c as part of wasm2c; the wasm2c tool generates code that wants to interface with the runtime API, and the upstream wasm-rt-impl.c is on the other side of that abstraction boundary and demonstrates one way to implement the API on a POSIXy or Win32 system. AFAIK this runtime implementation was written to be quite minimal and mostly just to get the spec tests running; I don't think it was trying to be re-entrant or thread-safe in a multi-threaded environment.

If this issue is really just about changes to wasm-rt-impl.c to make this runtime more sophisticated (which it looks like from your links in the issue?), I think this mostly comes down to questions about:

- "If we decide we want the example runtime to be re-entrant, do we think downstream users of the code will have consensus on the coordination mechanism used?" (E.g. mutexes or atomic data structures in `wasm_rt_register_func_type` and `wasm_rt_register_tag`?) As you know, I'm a little skeptical that any large user (e.g. Firefox or Chrome or whatever) is going to be happy with the type of lock or other thread-safety mechanism we choose, because they probably have their own way of doing this kind of thing.

- "How much complexity is worth it to make the example runtime thread-safe?" I see the example runtime as serving a valuable purpose in demonstrating the minimal needs to implement this API correctly (which of course we're using in implementing our own API). But I understand you see it as serving a valuable purpose in demonstrating something closer to what a sophisticated user would need. Do we want WABT to include two runtime implementations, one minimal and one intended to be closer to production-grade?

BTW we had a thread with longer-term vision stuff for wasm2c that you and your group might enjoy participating in to find areas of convergence or highlight and hopefully resolve disagreement in advance (https://github.com/WebAssembly/wabt/issues/2019).

**Comment by keithw** [comment]
> **Module specific state** Module specific state from the above would be the [g_func_types/g_func_type_count](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L64-L65)

I think the challenge is that this stuff isn't module-specific; the func types and tag types are shared across the whole WebAssembly store. This is necessary to allow one module to export a `funcref` which another module calls via `call_indirect`, or for one module to export an exception type (tag) which another module `throw`s (and maybe a third module `catch`es). Otherwise we wouldn't need the runtime global; the func/tag types could be part of the module's own state. :-/

You could imagine a crazy scheme where the func types are actually gigantic u32s that are some sort of hash of the function signature (which would avoid the need for any data structure at all!). Unfortunately I don't think this would work for tags because two tags with identical signatures are still distinct.

> **Thread specific state**
> 
> Other variables like [wasm_rt_jmp_buf](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L67) [wasm_rt_call_stack_depth/wasm_rt_saved_call_stack_depth](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L60-L61) [wasm exception related state](https://github.com/WebAssembly/wabt/blob/main/wasm2c/wasm-rt-impl.c#L69-L73) can be thought of as thread specific state that is part of the currently executing Wasm thread. We can thus create a new `struct wasm_rt_thread_state` for this data.
> 
> This struct is then passed in either
> 
>     1. as an extra function parameter to wasm2c functions (similar to the instance struct as of today) [See here](https://github.com/WebAssembly/wabt/compare/main...PLSysSec:wabt:remove-race-conditions-threadparam). This approach is clean but likely has some performance hit due to passing around extra parameters in every function
> 
>     2. stored in a thread local variable prior to execution [See here](https://github.com/WebAssembly/wabt/compare/main...PLSysSec:wabt:remove-race-conditions-tlsvar). This is probably more efficient, but the implementation must make sure to correctly set up the thread local variable prior to executing wasm2c functions, so it is more prone to error

Tentatively I think the thread-local approach might be my preference for how to deal with these -- maybe put all the runtime's state in one structure and then have a thread-local variable that points to an instance of the struct.

But another way to do it could be to augment the module instance structure to contain an opaque void* that is passed in at the time of instantiation, and then augment the runtime API functions to get the module instance pointer (just like all the other functions do). And that way, if the embedder wants to put the module instances in different "worlds" (e.g. one per thread), it can pass that parameter in at the time of instantiation, and it would be an error to try to link modules together that came from different worlds. And then the runtime API can use the void* to look up its runtime state. I don't feel that strongly either way...

**Comment by shravanrn** [comment]
Apologies for the delay. Had to deal with a paper submission the last week.

This below is an interesting question.

> Do we want WABT to include two runtime implementations, one minimal/demonstration and one intended to be closer to production-grade? Maybe this is a way to make everybody happy...

Sure, I don't have any concerns with this co-existing with more of a "tutorial" runtime (probably with a safety notice, that the tutorial runtime should not be used for production/security-critical use cases).

My main concern is that there needs to be a production ready runtime with the below criteria:

**Goals**:  Security. You should be able to use wasm2c in a "secure by default" manner (including the absence of race conditions) in your app with reasonable performance.

**Non-goals**: High performance in every production situation. To be clear, I will be striving to build a fast runtime and fast enough for browser use-cases, but I don't think we have to address every possible production need right away. I think it can be developed over time to steadily be improved to address needs that come up.

@sbc100 : thoughts on preferred approach (decouple prod vs tutorial, or migrate existing runtime to a production one?)

Re: some of the points you mention

> I think the challenge is that this stuff isn't module-specific; the func types and tag types are shared across the whole WebAssembly store. This is necessary to allow one module to return or export a funcref (or funcref-type table) that another module calls via call_indirect, or for one module to export an exception type (tag) that another module throws (and maybe a third module catches). Otherwise we wouldn't need the runtime globals; the type registry could be part of the module's own state. :-/

Ah, this was what I was afraid off. That is an unfortunate choice by the Wasm specification, as it introduces a bottleneck on anything that interacts with this global store. Given this requirement, locks may be the way to go as a starting point. We can investigate a more complex situation once we have a baseline solution.


> If we decide we want the example runtime to be re-entrant, do we think downstream users of the code will have consensus on the coordination mechanism used?

Probably not on the concensus front, but I think we should start by building a mechanism, and then as production uses come up up, we can optimize paths as needed / offer way to allow hosts to configure this.

> As you know, I was a little skeptical that any large user (e.g. Firefox or Chrome or whatever) is going to be happy with the type of lock or other thread-safety mechanism we would choose,

I can speak for the use of wasm2c in Firefox front. I think a simple approach with locks will probably be fine for now. We can investigate more complex solutions if it starts to become a performance bottleneck going forward.

> I wouldn't think a large C++ application will want to be calling setjmp on wasm_rt_jmp_buf before every call to a Wasm function -- they could prefer a wasm_rt_trap that throws a zero-cost-if-not-thrown C++ exception (which we're probably not going to do in the upstream runtime).

This isn't an issue for the moment. For Firefox, we have already disabled all of this prior to deployment (in our wasm2c fork). With the latest changes in mainline wasm2c, there are already enough knobs to turn these pieces. So we will similarly disable these pieces when we are ready to move away from the fork.

> "If we do make the upstream runtime re-entrant and start claiming it's thread-safe, how much do we care about ensuring that with tests?"

For the runtime pieces, I'm not sure there is much of a testing story right now (please correct me if I'm wrong). I think this speaks to a bigger question of "how do we test the runtime", that we I'll suggest we can address in a separate bug/discussion.

For any tests re the wasm generated code, I think this will probably be a part of the Wasm threading spec at some point, so I think it may make sense to wait for that to let that stabilize first, before we spend a lot of effort.

> BTW I had started a thread about longer-term vision stuff for wasm2c that you and your group might enjoy participating in to find areas of convergence or highlight and hopefully resolve disagreement in advance

Cool, thanks for sharing!

> Tentatively I think the thread-local approach might be my preference for how to deal with these ... But another way to do it could be to augment the module instance structure to contain an opaque void* that is passed in at the time of.

Yeah makes sense.
_Advantage of first approach:_ Better perf
_Advantage of second approach:_ Flexibility. (the first approach of creating and using a new variable for each host-thread can be thought of as a one possible implementation of the second technique)

I also don't feel strongly about this. I'll probably go with the first approach if if I need to change this, as we might as well get the perf wins, if there is no use-case for the flexibility at the moment.
